source/common/network/socket_interface_singleton.h

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include "envoy/network/socket.h"
+
+#include "common/singleton/threadsafe_singleton.h"
+
+namespace Envoy {
+namespace Network {
+
+SocketInterface* createDefaultSocketInterface();
+
+template <class T> class CustomInjectableSingleton {
```

Comment:
This looks, at first glance, like it does the same thing as `ThreadSafeSingleton` in `source/common/singleton/threadsafe_singleton.h`. If it's substantially different, please document exactly why it is needed, when it should be used, etc. And probably move it to the `singleton` directory.

Reply:
I did it this way for some of the reasons @lizan mentioned and because I wasn't really sure about the best approach :-)

The goal is to have an injectable and thread safe singleton with a default value. That is, if not configured, like is the case for the tests, SocketInterfaceImpl is used as a default implementation. Alternatively, it can be configured once per API init, to use a different implementation. 

I'll add more code comments once I clarify with @lizan if this approach is actually acceptable.